### PR TITLE
feat: add mesh enable config and process url in mesh mode

### PR DESCRIPTION
**Pull request by conghuhu**
## What is the purpose of the change

ApplicationConfig adds meshEnable configuration, ReferenceConfig adds providerPort configuration, and at the same time, api configuration, xml configuration, attribute configuration, and annotation configuration are modified.

After meshEnable is true, the registry is set to N/A, at least one tri protocol needs to be configured, and the ports configured by the default consumer and provider are the same. DubboReference only needs to configure providedBy (K8S Service name == provider application name) to call.

```properties
# dubbo-consumer.properties
dubbo.application.name=dubbo-samples-mesh-consumer
dubbo.registry.address=N/A
dubbo.protocol.name=tri
dubbo.protocol.port=50052
# enable mesh
dubbo.application.meshEnable=true
```
```java
// consumer
@DubboReference(providedBy = "dubbo-samples-mesh-provider")
private Greeter greeter;
```
If the provider and consumer service ports are inconsistent, you need to configure the port exposed by the provider in DubboReference, as follows:
```java
// consumer
@DubboReference(providedBy = "dubbo-samples-mesh-provider", providerPort = 50052)
private Greeter greeter;
```

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Review by wangchengming666 — APPROVED** [review]
test passed on my pc

File: dubbo-common/src/main/java/org/apache/dubbo/config/AbstractReferenceConfig.java [hunk 1]
```diff
@@ -82,6 +83,15 @@ public abstract class AbstractReferenceConfig extends AbstractInterfaceConfig {
...
     protected String providedBy;
 
+    /**
+     * The service port of the provider, the default is the triple port of the {@link ProtocolConfig#port}.
+     * In mesh mode: by default, dubbo will think that the ports of the consumer and the provider are the same.
+     * If the tri port of the consumer and the provider are inconsistent, please specify the tri port of provider on the consumer side
+     *
+     * @since 3.1.0
+     */
+    protected Integer providerPort;
```
Comment (CrazyHZM) [inline dubbo-common/src/main/java/org/apache/dubbo/config/AbstractReferenceConfig.java:93]: Need to consider the multi-port situation.

File: dubbo-common/src/main/java/org/apache/dubbo/config/ApplicationConfig.java [hunk 2]
```diff
@@ -203,6 +204,12 @@ public class ApplicationConfig extends AbstractConfig {
 
     private Boolean enableEmptyProtection;
 
+    /**
+     * enable mesh mode
+     * @since 3.1.0
+     */
+    private Boolean meshEnable;
```
Comment (CrazyHZM) [inline dubbo-common/src/main/java/org/apache/dubbo/config/ApplicationConfig.java:211]: `meshEnable` should be the behavior of the consumer and should be placed in the configuration of the consumer level

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/reference/ReferenceBeanBuilder.java [hunk 3]
```diff
@@ -198,6 +198,11 @@ public ReferenceBeanBuilder setProvidedBy(String providedBy) {
         return this;
     }
 
+    public ReferenceBeanBuilder setProviderPort(Integer providerPort) {
+        attributes.put(ReferenceAttributes.PROVIDER_PORT, providerPort);
+        return this;
+    }
+
```
Comment (CrazyHZM) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/reference/ReferenceBeanBuilder.java:205]: Remove useless code.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/reference/ReferenceAttributes.java [hunk 4]
```diff
@@ -52,6 +52,8 @@ public interface ReferenceAttributes {
 
     String PROVIDED_BY = "providedBy";
 
+    String PROVIDER_PORT = "providerPort";
+
```
Comment (CrazyHZM) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/reference/ReferenceAttributes.java:56]: Remove useless code.

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java [hunk 5]
```diff
@@ -439,6 +448,65 @@ private T createProxy(Map<String, String> referenceParameters) {
...
+                logger.warn("Can not get env variable: POD_NAMESPACE, it may not be running in the K8S environment , " +
+                        "finally use 'default' replace");
+            }
+            podNamespace = "default";
+        } else {
+            podNamespace = System.getenv("POD_NAMESPACE");
+        }
+
+        // url port, default 80.
+        AtomicInteger port = new AtomicInteger(DEFAULT_MESH_PORT);
```
Comment (CrazyHZM) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:481]: Why use `AtomicInteger`?

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/constants/CommonConstants.java [hunk 6]
```diff
@@ -546,4 +546,26 @@ public interface CommonConstants {
...
 
+    /**
+     * @since 3.1.0
+     */
+    String MESH_ENABLE = "mesh-enable";
+
+    /**
+     * @since 3.1.0
+     */
+    Integer DEFAULT_MESH_PORT = 80;
```
Comment (chickenlj) [inline dubbo-common/src/main/java/org/apache/dubbo/common/constants/CommonConstants.java:557]: Can we make the default mesh port the same as triple's default port? I think the default `80` port is not helpful for users using Triple, which is probably the most common case.

**Pull request merged**